crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/meta_net/tests.rs

```diff
@@ -0,0 +1,480 @@
+use super::*;
+
+use kitsune_p2p_fetch::OpHashSized;
+use kitsune_p2p_timestamp::Timestamp;
+use must_future::MustBoxFuture;
+use std::sync::Arc;
+
+use kitsune_p2p_types::{
+    bin_types::KitsuneSpace,
+    dependencies::lair_keystore_api,
+    dht::{
+        region::{Region, RegionCoords},
+        region_set::RegionSetLtcs,
+        spacetime::Topology,
+    },
+    dht_arc::DhtArcSet,
+    metrics::MetricRecord,
+    KOpData, KOpHash,
+};
+
+use crate::spawn::actor::FetchContext;
+use crate::spawn::actor::FetchKey;
+use crate::spawn::actor::FetchSource;
+use crate::spawn::actor::MaybeDelegate;
+use crate::spawn::actor::MetricExchangeMsg;
+use crate::spawn::actor::OpHashList;
+use crate::spawn::BroadcastData;
+
+use crate::event::GetAgentInfoSignedEvt;
+use crate::event::*;
+
+use crate::spawn::{Internal, InternalHandler, InternalHandlerResult};
+
+macro_rules! write_test_struct {
+    ($(
+        $ity:ty {
+            $(
+                fn $fna:ident (
+                    $fself:ty,
+                    $(
+                        $fpna:ident: $fpty:ty,
+                    )*
+                ) -> $fret1:ty, $fret2:ty $fdef:block
+            )*
+        }
+    )*) => {
+        pub struct Test {
+            recv: Arc<dyn Fn(MetaNetEvt) + 'static + Send + Sync>,
+            $($(
+                $fna: Arc<dyn Fn(
+                    $(
+                        $fpty,
+                    )*
+                ) -> $fret2 + 'static + Send + Sync>,
+            )*)*
+        }
+
+        impl std::fmt::Debug for Test {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                f.debug_struct("Test").finish()
+            }
+        }
+
+        impl Default for Test {
+            fn default() -> Self {
+                Self {
+                    recv: Arc::new(|_evt: MetaNetEvt| {}),
+                    $($(
+                        $fna: Arc::new(|$(
+                            $fpna: $fpty,
+                        )*| $fdef),
+                    )*)*
+                }
+            }
+        }
+
+        $(
+            impl $ity for RunningTest {
+                $(
+                    fn $fna(self: $fself, $(
+                        $fpna: $fpty,
+                    )*) -> $fret1 {
+                        (self.0.$fna)($(
+                            $fpna,
+                        )*).into()
+                    }
+                )*
+            }
+        )*
+    };
+}
+
+type HostRes<T> = Result<T, Box<dyn Send + Sync + std::error::Error>>;
+type HostRet<T> = std::pin::Pin<Box<dyn std::future::Future<Output = HostRes<T>> + 'static + Send>>;
+
+write_test_struct! {
+    KitsuneHost {
```

Comment:
There is a stub implementation for this in `crate::host_api::host_stub`, can that be used here?

Reply:
@ThetaSinner - This is set up with default handler callbacks, and the ability to replace them so the test can return whatever it wants or track the input parameters or whatever, more like a mock. So it's more versatile than the stub which must be the same across all tests using it.

Reply:
Makes sense, I'd like the other mocks/stubs to be similar at some point so maybe we can reconcile them at some point